File: item-list/src/main/java/com/gurunars/item_list/FetcherComplexPermutation.java
package com.gurunars.item_list;

import java.util.List;

import java8.util.function.BiFunction;


class FetcherComplexPermutation<ItemType extends Item> {

    private FetcherPermutationRange<ItemType> fetcher = new FetcherPermutationRange<>();

    public Change<ItemType> apply(
            List<ItemType> source,
            List<ItemType> target,
            int startOffset
    ) {
        FetcherPermutationRange.Range range = fetcher.apply(source, target);
        return new ChangeComplexPermutation<>(
                startOffset + range.getStart(),
                target.subList(range.getStart(), range.getEnd()));
    }
}

File: item-list/src/main/java/com/gurunars/item_list/FetcherComplexPermutation.kt
package com.gurunars.item_list


internal class FetcherComplexPermutation<ItemType : Item> {

    private val fetcher = FetcherPermutationRange<ItemType>()

    fun apply(
            source: List<ItemType>,
            target: List<ItemType>,
            startOffset: Int
    ): Change<ItemType> {
        val range = fetcher.apply(source, target)
        return ChangeComplexPermutation(
                startOffset + range.start,
                target.subList(range.start, range.end))
    }
}
